wasi/wasi_test.go

```diff
@@ -15,290 +15,298 @@ import (
+		require.EqualError(t, err, "arg[1] will exceed max buffer size 2")
+	})
+}
 
-// clockWat is a wasm module to call clock_time_get.
-//go:embed testdata/clock.wat
-var clockWat []byte
+func TestNewAPI_Args(t *testing.T) {
+	t.Run("valid", func(t *testing.T) {
+		o, err := Args("a", "bc")
+		require.NoError(t, err)
+		a := newAPI(o)
+		require.Equal(t, &nullTerminatedStrings{
+			nullTerminatedValues: [][]byte{
+				{'a', 0},
+				{'b', 'c', 0},
+			},
+			totalBufSize: 5,
+		}, a.args)
+	})
+	t.Run("error constructing args", func(t *testing.T) {
+		_, err := Args("\xff\xfe\xfd", "foo", "bar")
+		require.EqualError(t, err, "arg[0] is not a valid UTF-8 string")
+	})
+}
 
-func TestArgsAPISucceed(t *testing.T) {
+func TestApi_ArgsSizesGet(t *testing.T) {
 	ctx := context.Background()
-	tests := []struct {
-		name            string
-		args            []string
-		expectedArgs    [][]byte
-		expectedBufSize uint32
-	}{
-		{
-			name:            "no args",
-			args:            nil,
-			expectedArgs:    [][]byte{},
-			expectedBufSize: 0,
-		},
-		{
-			name:            "empty",
-			args:            []string{},
-			expectedArgs:    [][]byte{},
-			expectedBufSize: 0,
-		},
-		{
-			name: "simple",
-			args: []string{"foo", "bar", "foobar", "", "baz"},
-			expectedArgs: [][]byte{
-				[]byte("foo\x00"),
-				[]byte("bar\x00"),
-				[]byte("foobar\x00"),
-				[]byte("\x00"),
-				[]byte("baz\x00"),
-			},
-			expectedBufSize: 20,
-		},
-		{
-			name: "utf-8 string",
-			// "😨", "🤣", and "️🏃‍♀️" have 4, 4, and 13 bytes respectively
-			args: []string{"😨🤣🏃\u200d♀️", "foo", "bar"},
-			expectedArgs: [][]byte{
-				[]byte("😨🤣🏃\u200d♀️\x00"),
-				[]byte("foo\x00"),
-				[]byte("bar\x00"),
-			},
-			expectedBufSize: 30,
-		},
-		{
-			name: "invalid utf-8 string",
-			args: []string{"\xff\xfe\xfd", "foo", "bar"},
-			expectedArgs: [][]byte{
-				[]byte("\xff\xfe\xfd\x00"),
-				[]byte("foo\x00"),
-				[]byte("bar\x00"),
-			},
-			expectedBufSize: 12,
-		},
-	}
+	args, err := Args("a", "bc")
+	require.NoError(t, err)
 
-	for _, tt := range tests {
-		tc := tt
+	argc := uint32(1)        // arbitrary offset
+	argvBufSize := uint32(6) // arbitrary offset
+	maskLength := 11         // number of bytes to write '?' to tell what we've written
+	expectedMemory := []byte{
+		'?',                // argc is after this
+		0x2, 0x0, 0x0, 0x0, // little endian-encoded arg count
+		'?',                // argvBufSize is after this
+		0x5, 0x0, 0x0, 0x0, // little endian-encoded size of null terminated strings
+		'?', // stopped after encoding
+	} // tr
+	store, wasiAPI := instantiateWasmStore(t, argsWat, "test", args)
+
+	t.Run("API.ArgsSizesGet", func(t *testing.T) {
+		maskMemory(store, maskLength)
+
+		// provide a host context we call directly
+		hContext := wasm.NewHostFunctionCallContext(ctx, store.Memories[0])
+
+		// invoke ArgsSizesGet directly and check the memory side-effects!
+		errno := wasiAPI.ArgsSizesGet(hContext, argc, argvBufSize)
```

Comment:
this is pretty cool imho as we know nothing in the re-exporting stuff can interfere with it. this will pass even if the other doesn't due to some odd reason